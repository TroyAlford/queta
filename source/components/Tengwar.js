import React, { Component, Fragment } from 'react'
import Glaemscribe from '../../vendor/build/glaemscribe'
import './Tengwar.scss'

Glaemscribe.resource_manager.load_modes()
Glaemscribe.resource_manager.load_charsets()
if (!window.Glaemscribe) window.Glaemscribe = Glaemscribe;

const modes = Glaemscribe.resource_manager.loaded_modes
export const MODES = Object.values(modes)
  .map(mode => ({
    charsets: mode.supported_charsets,
    humanName: mode.human_name,
    language: mode.language,
    languageClass: `language-${mode.name}`,
    name: mode.name,
    writing: mode.writing,
    writingClass: `writing-${mode.writing.toLowerCase()}`,

    transcriber: modes[mode.name]
  }))

window.MODES = MODES;

export default class Tengwar extends Component {
  static defaultProps = {
    bold: false,
    italic: false,
    language: 'quenya',
    writing: 'Runes',
    tagName: 'span',
    typeface: 'elfica',
  }

  get language() {
    return (
      MODES.find(m => m.name === this.props.language && m.writing === this.props.writing) ||
      MODES.find(m => m.name === this.props.language) ||
      MODES[0]
    )
  }
  get charset() {
    return (
      this.language.charsets[this.props.charset] ||
      Object.values(this.language.charsets)[0]
    )
  }

  transcribe = (text) => this.language.transcriber.transcribe(text, this.charset)
  transcribeChildren = nodes => {
    const children = Array.isArray(nodes) ? nodes : [nodes]

    return children.map((child, index) => {
      if (typeof child === 'string') {
        const [success, transcription] = this.transcribe(child)
        if (success) return transcription
      }

      if (child && child.props && child.props.children) {
        const grandChildren = child.props.children
        const Type = child.type
        return (
          <Type key={index} {...child.props}>
            {this.transcribeChildren(Array.isArray(grandChildren) ? grandChildren : [grandChildren])}
          </Type>
        )
      }

      return child
    })
  }
  render = () => {
    const { bold, italic, children, tagName: Tag, typeface } = this.props
    const language = this.language
    const classes = [
      'glaemscribe',
      language.languageClass,
      language.writingClass,
      `typeface-${typeface}`,
      bold && 'bold',
      italic && 'italic',
    ].filter(Boolean)

    return (<Tag className={classes.join(' ')}>{this.transcribeChildren(children)}</Tag>)
  }
}
