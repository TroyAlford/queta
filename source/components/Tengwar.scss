.glaemscribe {
	line-height: 1.5em;

	&.loading::after {
		content: 'Loading...';
		font-style: italic;
		opacity: .75;
	}
}
