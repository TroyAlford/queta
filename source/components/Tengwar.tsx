import type * as Glaemscribe from 'glaemscribe'
import React, { Component } from 'react'

type TProps = {
	className?: string,
	language?: string,
	typeface?: string,
	writing?: Glaemscribe.Writing,
}

type TState = {
	loading: boolean,
}

export class Tengwar extends Component<TProps, TState> {
	private static dependencies = Promise.all([
		import('../../temp/glaemscribe.built.js').then(module => {
			const processor = module.default as Glaemscribe.default

			const manager = processor.resource_manager
			manager.load_charsets()
			manager.load_modes()
			Tengwar.charsets = Array.from(Object.values(manager.loaded_charsets))
			Tengwar.modes = Array.from(Object.values(manager.loaded_modes))

			return processor
		}),
		import('./Tengwar.scss'),
		import('../../temp/glaemscribe.built.scss'),
	])

	public static defaultProps: TProps = {
		language: 'quenya-tengwar-classical',
		typeface: 'tengwar_guni_sindarin',
		writing: 'Tengwar',
	}
	public static modes: Glaemscribe.Mode[] = []
	public static charsets: Glaemscribe.Charset[] = []

	public state: TState = {
		loading: !window.Glaemscribe,
	}

	componentDidMount = (): void => {
		if (this.state.loading) {
			Tengwar.dependencies.then(() => this.setState({ loading: false }))
		}
	}

	get charset(): Glaemscribe.Charset | undefined {
		const { mode } = this
		if (!mode) return undefined

		const charsets = mode.supported_charsets || {}
		return charsets[this.props.typeface] ?? mode.default_charset
	}
	get mode(): Glaemscribe.Mode | undefined {
		return Tengwar.modes.find(m => m.name === this.props.language)
	}
	get typeface(): string { return this.charset?.name }

	#renderChildren = (children: React.ReactNode): React.ReactChild[] => (
		(React.Children.toArray(children) as React.ReactChild[]).map((child, index) => {
			if (typeof child === 'string') {
				const [success, transcription] = this.#renderText(child)
				if (success) { return transcription }
			} else if (typeof child === 'number') {
				return child
			} else if (child?.props?.children) {
				const grandChildren = child.props.children
				const Type = child.type
				return (
					<Type key={index} {...child.props}>
						{this.#renderChildren(
							Array.isArray(grandChildren)
								? grandChildren
								: [grandChildren],
						)}
					</Type>
				)
			}

			return child
		})
	)
	#renderText = (text: string): string => this.mode?.transcribe(text, this.charset) ?? text

	render = (): JSX.Element => {
		const { children, className } = this.props
		const { loading } = this.state
		const classes = [
			'glaemscribe',
			className,
			...(loading ? ['loading'] : [this.typeface, this.mode?.name]),
		].filter(Boolean).join(' ')

		if (loading) { return <div className={classes} /> }

		return (
			<div className={classes}>
				{(
					this.mode
						? this.#renderChildren(children)
						: children
				)}
			</div>
		)
	}
}

window.Tengwar = Tengwar
