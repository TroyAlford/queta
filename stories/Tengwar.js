import React from 'react'
import { storiesOf } from '@storybook/react'
import { withKnobs, select, text } from '@storybook/addon-knobs/react'
import Tengwar, { MODES } from '../source/components/Tengwar'

const stories = storiesOf('Tengwar', module)

stories.addDecorator(withKnobs)

const LANGUAGES = MODES.reduce((map, mode) => ({
  ...map,
  [mode.name]: mode.humanName
}), {})
const TYPEFACE_OPTIONS = {
  annatar: 'Tengwar Annatar',
  eldamar: 'Tengwar Eldamar',
  elfica: 'Tengwar Elfica',
  parmaite: 'Tengwar Parmaite',
  sindarin: 'Tengwar Sindarin',
}
const WRITING_OPTIONS = {
  cirth: 'Cirth',
  gothic: 'Gothic',
  runes: 'Runes',
  sarati: 'Sarati',
  tengwar: 'Tengwar',
}

const DEFAULT_TEXT = `
Ai ! laurië lantar lassi súrinen ,
yéni únótimë ve rámar aldaron !
Yéni ve lintë yuldar avánier
mi oromardi lissë-miruvóreva
Andúnë pella , Vardo tellumar
nu luini yassen tintilar i eleni
ómaryo airetári-lírinen .

Sí man i yulma nin enquantuva ?
`

stories.add('Full Example', () => (
  <Tengwar
    children={text('Text', DEFAULT_TEXT, 'foo')}
    language={select('Language', LANGUAGES, 'Quenya', 'foo')}
    typeface={select('Typeface', TYPEFACE_OPTIONS, 'Tengwar Elfica', 'foo')}
    writing={select('Writing', WRITING_OPTIONS, 'Tengwar', 'foo')}
    options={{
      always_use_romen_for_r: true,
      split_diphthongs: true,
    }}
  />
))
