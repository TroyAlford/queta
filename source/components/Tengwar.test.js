import React from 'react'
import Tengwar from './Tengwar'
import { shallow } from 'enzyme'

const classString = (language, writing, typeface) => (
  `glaemscribe language-${language} writing-${writing} typeface-${typeface}`
)

describe('tengwar', () => {
  it('defaults to quenya|tengwar|elfica', () => {
    const wrapper = shallow(<Tengwar />)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('transpiles single child string contents', () => {
    const wrapper = shallow(<Tengwar>Minas Tirith</Tengwar>)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('transpiles nested children & their contents', () => {
    const wrapper = shallow(
      <Tengwar>
        Minas Tirith
        <i>Ost Valaron</i>
        <b>Anduin</b>
        Minas Valaron
      </Tengwar>
    )
    expect(wrapper.html()).toMatchSnapshot()
  })
})
