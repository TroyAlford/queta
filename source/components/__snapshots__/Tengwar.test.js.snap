// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tengwar defaults to quenya|tengwar|elfica 1`] = `"<span class=\\"glaemscribe language-quenya writing-tengwar typeface-elfica\\"></span>"`;

exports[`tengwar transpiles nested children & their contents 1`] = `"<span class=\\"glaemscribe language-quenya writing-tengwar typeface-elfica\\">t%5#8 1T7T3<i>\`N81 yEjE7Y5</i><b>\`C2lU5</b>t%5#8 yEjE7Y5</span>"`;

exports[`tengwar transpiles single child string contents 1`] = `"<span class=\\"glaemscribe language-quenya writing-tengwar typeface-elfica\\">t%5#8 1T7T3</span>"`;
